File: PlayCover/AppInstaller/optool/headers.m
//
//  headers.m
//  optool
//  Copyright (c) 2014, Alex Zielenski
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  * Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
//  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
//  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
//  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
//  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "headers.h"
#import <mach-o/loader.h>
#import <mach-o/fat.h>
#import "NSData+Reading.h"

struct thin_header headerAtOffset(NSData *binary, uint32_t offset) {
    struct thin_header macho;
    macho.offset = offset;
    macho.header = *(struct mach_header *)(binary.bytes + offset);
    macho.size = sizeof(struct mach_header_64);

    if (macho.header.cputype != CPU_TYPE_X86_64 && macho.header.cputype != CPU_TYPE_I386 && macho.header.cputype != CPU_TYPE_ARM && macho.header.cputype != CPU_TYPE_ARM64){
        macho.size = 0;
    }

    return macho;
}

struct thin_header *headersFromBinary(struct thin_header *headers, NSData *binary, uint32_t *amount) {
    // In a MachO/FAT binary the first 4 bytes is a magic number
    // which gives details about the type of binary it is
    // CIGAM and co. mean the target binary has a byte order
    // in reverse relation to the host machine so we have to swap the bytes
    uint32_t magic = [binary intAtOffset:0];
    uint32_t numArchs = 0;

    // The binary is thin, meaning it contains only one architecture
    if (magic == MH_MAGIC_64) {
        struct thin_header macho = headerAtOffset(binary, 0);
        if (macho.size > 0) {
            LOG("Found thin header...");

            numArchs++;
            headers[0] = macho;
        }
    } else {
        LOG("No headers found.");
    }
    
    *amount = numArchs;
    return headers;
}

File: PlayCover/AppInstaller/optool/Headers.swift
//
//  Headers.swift
//  PlayCover
//
//  Created by Isaac Marovitz on 20/09/2022.
//

import Foundation

// swiftlint:disable type_name
public struct thin_header {
    var offset: UInt32 = 0
    var size: UInt32 = 0
    var header: mach_header = mach_header()
}

class Headers {
    public static func headerAtOffset(binary: inout OffsetNSData, offset: UInt32) -> thin_header {
        var macho = thin_header()
        macho.offset = offset
        macho.header = (binary.data.bytes + Int(offset)).load(as: mach_header.self)
        macho.size = UInt32(MemoryLayout.stride(ofValue: mach_header_64.self))

        if macho.header.cputype != CPU_TYPE_ARM64 {
            macho.size = 0
        }

        return macho
    }

    public static func headersFromBinary(_ headers: inout [thin_header],
                                         _ binary: inout OffsetNSData,
                                         _ amount: inout UInt32) {
        let magic: UInt32 = binary.intAtOffset(offset: 0)
        var numArchs: UInt32 = 0

        if magic == MH_MAGIC_64 {
            let macho = headerAtOffset(binary: &binary, offset: 0)
            if macho.size > 0 {
                print("Found thin header...")

                numArchs += 1
                headers[0] = macho
            }
        } else {
            print("No headers found.")
        }

        amount = numArchs
    }
}
